### ISSUE TITLE
immutable.HashMap::get[OrElse] performance regression in Scala 2.12

### ISSUE THREAD

**Issue by fwbrasil** [scala/bug#11172]
We're currently migrating from Scala 2.11.11 to 2.12.6 at Twitter and observed a considerable performance regression. One of the factors seems to be more costly immutable hash map lookups. I've copied [HashMapBenchmark](https://github.com/scala/scala/blob/2.13.x/test/benchmarks/src/main/scala/scala/collection/immutable/HashMapBenchmark.scala) and compared the results:

Scala 2.11.11
```
[info] Benchmark                            (size)  (stringsOnly)  (useMissingValues)  Mode  Cnt      Score      Error  Units
[info] ImmutableHashMapBenchmark.get            10          false                true  avgt    5    225.436 ±   10.088  ns/op
[info] ImmutableHashMapBenchmark.get           100          false                true  avgt    5   3321.597 ±  180.465  ns/op
[info] ImmutableHashMapBenchmark.get          1000          false                true  avgt    5  50101.150 ± 6392.813  ns/op
[info] ImmutableHashMapBenchmark.getOrElse      10          false                true  avgt    5    287.946 ±   15.273  ns/op
[info] ImmutableHashMapBenchmark.getOrElse     100          false                true  avgt    5   3533.484 ±   55.820  ns/op
[info] ImmutableHashMapBenchmark.getOrElse    1000          false                true  avgt    5  54772.205 ± 3833.055  ns/op
```

Scala 2.12.6
```
[info] Benchmark                            (size)  (stringsOnly)  (useMissingValues)  Mode  Cnt      Score      Error  Units
[info] ImmutableHashMapBenchmark.get            10          false                true  avgt    5    271.087 ±    8.198  ns/op
[info] ImmutableHashMapBenchmark.get           100          false                true  avgt    5   3297.224 ±   32.886  ns/op
[info] ImmutableHashMapBenchmark.get          1000          false                true  avgt    5  64384.090 ± 1776.272  ns/op
[info] ImmutableHashMapBenchmark.getOrElse      10          false                true  avgt    5    303.964 ±    4.345  ns/op
[info] ImmutableHashMapBenchmark.getOrElse     100          false                true  avgt    5   4688.946 ±  314.729  ns/op
[info] ImmutableHashMapBenchmark.getOrElse    1000          false                true  avgt    5  64319.199 ± 1338.750  ns/op
```

To reproduce:
```
git clone git@github.com:fwbrasil/scala-graal.git
cd scala-graal
sbt "+jmh:run -w 5 -r 5 -f 1 -wi 5 -i 5 -gc true ImmutableHashMapBenchmark.get"
```
Notes: 
- the other method benchmarks (`contains` and `updated`) actually have better performance on 2.12.6
- the project is called scala-graal because I've been using it to test graal changes, but the jmh benchmark runs on C2 by default.

**Comment by retronym** [comment]
The gross regression is fixed by scala/scala#7256

```
$ (for V in 2.11.11 2.12.6 2.12.7-bin-d96b8eb-SNAPSHOT; do echo "SCALA=$V"; sbt "++$V" "jmh:run -w 1 -r 1 ImmutableHashMapBenchmark.get -psize=1000  -f4"; done) | egrep 'SCALA=|ImmutableHashMapBenchmark.*ns/op'
SCALA=2.11.11
[info] ImmutableHashMapBenchmark.get          1000          false                true  avgt   40  35258.452 ±  243.577  ns/op
[info] ImmutableHashMapBenchmark.getOrElse    1000          false                true  avgt   40  46796.782 ± 1567.576  ns/op
SCALA=2.12.6
[info] ImmutableHashMapBenchmark.get          1000          false                true  avgt   40  48209.879 ±  149.473  ns/op
[info] ImmutableHashMapBenchmark.getOrElse    1000          false                true  avgt   40  53771.235 ± 1075.223  ns/op
SCALA=2.12.7-bin-d96b8eb-SNAPSHOT
[info] ImmutableHashMapBenchmark.get          1000          false                true  avgt   40  38185.885 ± 551.126  ns/op
[info] ImmutableHashMapBenchmark.getOrElse    1000          false                true  avgt   40  44612.994 ± 736.052  ns/op
```

There is still a small regression which might be explained by the trait encoding changes, and the vaguaries of C2's choice of inlining boundaries, but hopefully this gets us close enough to mean that real-applications won't be affected.

**Comment by ShaneDelmore** [comment]
Thanks!  I'll be taking 2.12.7 for a spin today.
